### PR TITLE
Add completions stream API

### PR DESCRIPTION
**Pull request by srnagar**

### CODE REVIEW

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/GetCompletionsStreamSample.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            .endpoint(endpoint)
+            .credential(new AzureKeyCredential(azureOpenaiKey))
+            .buildClient();
+
+        List<String> prompt = new ArrayList<>();
+        prompt.add("Say this is a test");
+        IterableStream<Completions> completionsStream = client.getCompletionsStream(deploymentOrModelId,
+            new CompletionsOptions(prompt).setMaxTokens(1000).setStream(true));
+
+        completionsStream.forEach(completions -> {
```
Comment (jpalvarezl) [inline sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/GetCompletionsStreamSample.java:43]: Is the lambda passed to `forEach` subscribing to the `Flux` ? Meaning, that if we were to translate this into a test, would we be able to embed assertions within it?

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/implementation/OpenAIServerSentEvents.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+                currentLine = lastLine.get() + currentLine;
+                List<T> values = new ArrayList<>();
+                while (currentLine != null) {
+                    if (expectEmptyLine.get() && !currentLine.isEmpty()) {
+                        return Flux.error(new UnsupportedOperationException("Multi-line data not supported"));
+                    }
+
+                    if (!expectEmptyLine.get()) {
+                        expectEmptyLine.set(true);
+                        String[] split = currentLine.split(":", 2);
```
Comment (jpalvarezl) [inline sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/implementation/OpenAIServerSentEvents.java:48]: I believe the stream also sends a `[DONE]` message when it is in fact done. Should we check for that? I am viscerally opposed to the idea, but can't quite justify why. Is it not a standard way of finishing a stream?

**Review by jpalvarezl — APPROVED** [review]

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/AsyncRestProxy.java [hunk 3]
```diff
@@ -176,12 +178,17 @@ static Mono<?> handleBodyReturnType(HttpResponse sourceResponse, Function<byte[]
             // Mono<Flux<ByteBuffer>>
             asyncResult = Mono.just(sourceResponse.getBody());
         } else if (TypeUtil.isTypeOrSubTypeOf(entityType, BinaryData.class)) {
+            HttpHeader httpHeader = sourceResponse.getHeaders().get(HttpHeaderName.fromString("Content-Type"));
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/AsyncRestProxy.java:181]: nit: We have `HttpHeaderName.CONTENT_TYPE`

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/AsyncRestProxy.java [hunk 4]
```diff
@@ -176,12 +178,17 @@ static Mono<?> handleBodyReturnType(HttpResponse sourceResponse, Function<byte[]
...
             // Mono<BinaryData>
             // The raw response is directly used to create an instance of BinaryData which then provides
             // different methods to read the response. The reading of the response is delayed until BinaryData
             // is read and depending on which format the content is converted into, the response is not necessarily
             // fully copied into memory resulting in lesser overall memory usage.
-            asyncResult = BinaryData.fromFlux(sourceResponse.getBody());
+            if (httpHeader != null && httpHeader.getValue() != null && httpHeader.getValue().equals("text/event-stream")) {
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/AsyncRestProxy.java:187]: nit: Flip the constant to the left-hand side of the `.equals`

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/SyncRestProxy.java [hunk 5]
```diff
@@ -189,7 +191,12 @@ private Object handleBodyReturnType(HttpResponseDecoder.HttpDecodedResponse resp
             // different methods to read the response. The reading of the response is delayed until BinaryData
             // is read and depending on which format the content is converted into, the response is not necessarily
             // fully copied into memory resulting in lesser overall memory usage.
-            result = response.getSourceResponse().getBodyAsBinaryData();
+            HttpHeader httpHeader = response.getSourceResponse().getHeaders().get(HttpHeaderName.fromString("Content-Type"));
+            if (httpHeader != null && httpHeader.getValue() != null && httpHeader.getValue().equals("text/event-stream")) {
```
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/SyncRestProxy.java:195]: Same comments

File: sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/GetCompletionsStreamSample.java (hunk 1 above)
Comment (srnagar) [reply sdk/openai/azure-ai-openai/src/samples/java/com/azure/ai/openai/GetCompletionsStreamSample.java:43]: Yes, forEach() will subscribe to the underlying flux and yes, you can have assertions in this lambda for testing.

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/implementation/OpenAIServerSentEvents.java (hunk 2 above)
Comment (srnagar) [reply sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/implementation/OpenAIServerSentEvents.java:52]: Added a check to ensure the stream ends gracefully.

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/AsyncRestProxy.java (hunk 3 above)
Comment (srnagar) [reply sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/AsyncRestProxy.java]: Done

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/AsyncRestProxy.java (hunk 4 above)
Comment (srnagar) [reply sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/AsyncRestProxy.java]: Done

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/SyncRestProxy.java (hunk 5 above)
Comment (srnagar) [reply sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/SyncRestProxy.java]: Done

**Pull request merged**